mmv1/third_party/terraform/functions/zone_from_id_test.go

```diff
@@ -0,0 +1,116 @@
+package functions_test
+
+import (
+	"fmt"
+	"regexp"
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/hashicorp/terraform-provider-google/google/acctest"
+	"github.com/hashicorp/terraform-provider-google/google/envvar"
+)
+
+func TestAccProviderFunction_zone_from_id(t *testing.T) {
+	t.Parallel()
+	// Skipping due to requiring TF 1.8.0 in VCR systems : https://github.com/hashicorp/terraform-provider-google/issues/17451
+	acctest.SkipIfVcr(t)
+
+	zone := envvar.GetTestZoneFromEnv()
+	zoneRegex := regexp.MustCompile(fmt.Sprintf("^%s$", zone))
+
+	context := map[string]interface{}{
+		"function_name": "zone_from_id",
+		"output_name":   "zone",
+		"resource_name": fmt.Sprintf("tf-test-zone-id-func-%s", acctest.RandString(t, 10)),
+	}
+
+	acctest.VcrTest(t, resource.TestCase{
+		ProtoV5ProviderFactories: acctest.ProtoV5ProviderFactories(t),
+		Steps: []resource.TestStep{
+			{
+				// Can get the zone from a resource's id in one step
+				// Uses google_compute_network_endpoint_group resource's id attribute with format projects/{{project}}/zones/{{zone}}/networkEndpointGroups/{{name}}
+				Config: testProviderFunction_get_zone_from_resource_id(context),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestMatchOutput(context["output_name"].(string), zoneRegex),
+				),
+			},
+			{
+				// Can get the zone from a resource's self_link in one step
+				// Uses google_compute_network_endpoint_group resource's self_link attribute
+				Config: testProviderFunction_get_zone_from_resource_self_link(context),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestMatchOutput(context["output_name"].(string), zoneRegex),
+				),
+			},
+		},
+	})
+}
+
+func testProviderFunction_get_zone_from_resource_id(context map[string]interface{}) string {
+	return acctest.Nprintf(`
+# terraform block required for provider function to be found
+terraform {
+  required_providers {
+    google = {
+      source = "hashicorp/google"
+    }
+  }
+}
+
+data "google_compute_network" "default" {
+  name = "default"
+}
+
+resource "google_compute_subnetwork" "default" {
+  name          = "%{resource_name}"
+  ip_cidr_range = "10.2.0.0/16"
```

Comment:
CIDRs will cause an issue, so this test should be refactored with a [persistent disk](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/compute_disk) as a better zonal resource example versus `google_compute_network_endpoint_group` below.

Reply:
Same change should be made in `testProviderFunction_get_zone_from_resource_self_link`

@c2thorn I'm holding off pushing a new commit with this to avoid blocking review, but this is part of my self review

Reply:
Addressed in [8a81243](https://github.com/GoogleCloudPlatform/magic-modules/pull/10060/commits/8a81243db5a8cb09b2962b400c5c3ab8dc692cbd) and [1f17ec0](https://github.com/GoogleCloudPlatform/magic-modules/pull/10060/commits/1f17ec0863fd262c251e5152b30e5b6036758b70)